File: java/gradientBitmapCalculator.java
class gradientBitmapCalculator{

    public static void fillGradientLine(int[] arr, int line, int w, int a, int b){
        int delta = b - a;
        int offset = line * w;
        for (int i = 0; i < w; i++){
            float fraction = (float) i / (float) w;
            arr[offset + i] = a + (int) ((float) delta * fraction);
        }
    }
    
    public static int[] generateGrayGradient(int h, int w, int a, int b){
        int size = w * h;
        int[] arr = new int[size];
        for (int i = 0; i < h; i++){
            fillGradientLine(arr, i, w, a, b);
        }
        return arr;
    }
    public static void saveGradient(int[] arr, int h, int w, int b){
        System.out.printf(java.lang.String.format("P2\n%d %d\n%d\n", w, h, b));
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++){
                System.out.print(arr[i * w + j]);
                System.out.print(" ");
            }
            System.out.println("");
        }
    }
    
    public static void main(String[] args) {
        int h = 1000;
        int w = 1000;
        int a = 55;
        int b = 233;

        final double startTime = System.nanoTime() / 1000000.0;
        generateGrayGradient(h, w, a, b);
        final double endTime = System.nanoTime() / 1000000.0;
        System.out.println(endTime - startTime);

        // saveGradient(arr, h, w, b);

        
    }
}

File: kt/gradientBitmapCalculator.kt
@kotlin.ExperimentalUnsignedTypes
fun fillGradientLine(arr: kotlin.UIntArray, line: Int, w: Int, a: UByte, b: UByte): Unit {
    val delta = b - a
    val offset = line * w
    for (i in 0 until w) {
        var fraction = i.toFloat() / w.toFloat()
        arr[offset + i] = a.toUByte() + (delta.toFloat() * fraction).toUInt()
    }
}

@kotlin.ExperimentalUnsignedTypes
fun generateGrayGradient(h: Int, w: Int, a: UByte, b: UByte): kotlin.UIntArray{
    val size = w * h
    val arr = kotlin.UIntArray(size)
    for (i in 0 until h){
        fillGradientLine(arr, i, w, a, b)
    }
    return arr
}

@kotlin.ExperimentalUnsignedTypes
fun saveGradient(arr: kotlin.UIntArray, h: Int, w: Int, b:UByte): Unit {
    print(java.lang.String.format("P2\n%d %d\n%d\n", w, h.toInt(), b.toInt()))
    for (i in 0 until h) {
        for (j in 0 until w){
            print(arr[i * w + j])
            print(" ")
        }
        println("")
    }
}

@kotlin.ExperimentalUnsignedTypes
fun main() {
    val h = 50
    val w = 50
    val a: UByte = 55.toUByte()
    val b: UByte = 233.toUByte()

    val start = System.nanoTime();
    generateGrayGradient(h, w, a, b)
    val timeMilli = (System.nanoTime() - start) / 1000000.0f
    println("$timeMilli")
    //saveGradient(arr, h, w, b)
}
